service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpConnectionHandler.java

```diff
@@ -0,0 +1,487 @@
+                }
+            }
+        }
+        else //Sender link received an ack for sent message
+        {
+            log.trace("Received acknowledgement for a message this process sent");
+            Delivery d = event.getDelivery();
+
+            DeliveryState remoteState = d.getRemoteState();
+
+            onMessageAcknowledged(remoteState);
+
+            d.settle();
+        }
+    }
+
+    @Override
+    public void onConnectionBound(Event event)
+    {
+        Transport transport = event.getConnection().getTransport();
+        if (transport != null)
+        {
+            if (this.iotHubServiceClientProtocol == IotHubServiceClientProtocol.AMQPS_WS)
+            {
+                log.debug("Adding websocket layer to transport");
+                WebSocketImpl webSocket = new WebSocketImpl();
+                webSocket.configure(this.webSocketHostName, WEBSOCKET_PATH, 0, WEBSOCKET_SUB_PROTOCOL, null, null);
+                ((TransportInternal)transport).addTransportLayer(webSocket);
+            }
+
+            Sasl sasl = transport.sasl();
+            sasl.plain(this.userName, this.sasToken);
+
+            SslDomain domain = makeDomain(SslDomain.Mode.CLIENT);
+            domain.setPeerAuthentication(SslDomain.VerifyMode.VERIFY_PEER);
+            transport.ssl(domain);
+        }
+    }
+
+    @Override
+    public void onConnectionInit(Event event)
+    {
+        this.connection = event.getConnection();
+        this.connection.setHostname(hostName);
+        Session ssn = this.connection.session();
+
+        Map<Symbol, Object> properties = new HashMap<>();
+        properties.put(Symbol.getSymbol(TransportUtils.versionIdentifierKey), TransportUtils.USER_AGENT_STRING);
+
+        log.trace("Opening connection");
+        this.connection.open();
+
+        log.trace("Opening session");
+        ssn.open();
+        openLinks(ssn, properties);
+    }
+
+    @Override
+    public void onLinkInit(Event event)
+    {
+        Link link = event.getLink();
+        if (event.getLink().getName().equals(tag))
+        {
+            Target t = new Target();
+            t.setAddress(endpoint);
+            Source source = new Source();
+            source.setAddress(endpoint);
+            link.setTarget(t);
+            link.setSource(source);
+        }
+    }
+
+    @Override
+    public void onLinkRemoteOpen(Event event)
+    {
+        log.debug("Link with name {} opened remotely", event.getLink().getName());
+        openLinksLatch.countDown();
+    }
+
+    @Override
+    public void onConnectionRemoteOpen(Event event)
+    {
+        log.debug("Connection opened remotely");
+    }
+
+    @Override
+    public void onSessionRemoteOpen(Event event)
+    {
+        log.debug("Session opened remotely");
+    }
+
+    @Override
+    public void onConnectionRemoteClose(Event event)
+    {
+        super.onConnectionRemoteClose(event);
+        event.getTransport().close_tail();
+        releaseOpenLatch();
```

Comment:
do we need to clean up reactor reference as well?

Reply:
For all cases where transport error occurs, or if connection/session/link is closed remotely, we respond by starting to close the local links/sessions/connections. At the end of that process, the reactor gets stopped. After the reactor stops, the reactor runner thread will free the reactor's resources. I can talk you through this bit, because it is a bit tricky to follow